runtime/src/status_cache.rs

```diff
@@ -122,186 +162,361 @@ impl<T: Serialize + Clone> StatusCache<T> {
 
-        let forks = hash_map.2.entry(key_slice).or_default();
-        forks.push((slot, res.clone()));
+            let mut forks = hash_map.entry(key_slice).or_default();
+            forks.push((slot, res.clone()));
+        }
 
-        self.add_to_slot_delta(transaction_blockhash, slot, key_index, key_slice, res);
+        self.add_to_slot_delta(blockhash, slot, key_index, key_slice, res);
     }
 
-    // Add this key slice to the list of key slices for this slot and blockhash
-    // combo.
+    // Add this key slice to the list of key slices for this slot and blockhash combo.
     fn add_to_slot_delta(
-        &mut self,
-        transaction_blockhash: &Hash,
+        &self,
+        blockhash: &Hash,
         slot: Slot,
         key_index: usize,
         key_slice: [u8; CACHED_KEY_SIZE],
         res: T,
     ) {
-        let mut fork_entry = self.slot_deltas.entry(slot).or_default().lock().unwrap();
-        let (_key_index, hash_entry) = fork_entry
-            .entry(*transaction_blockhash)
-            .or_insert((key_index, vec![]));
-        hash_entry.push((key_slice, res))
+        let fork_entry = self.slot_deltas.get_or_insert_with(&slot, || {
+            DashMap::with_hasher_and_shard_amount(AHashRandomState::default(), KEY_SHARDS)
+        });
+
+        // In the vast majority of the cases, there will already be an entry for this blockhash, so
+        // do a get() first so that we can avoid taking a write lock on the corresponding shard.
+        if let Some(fork_entry) = fork_entry.get(blockhash) {
+            let (_key_index, txs) = &*fork_entry;
+            // txs is a ConcurrentVec so we can push without any locking
+            txs.push((key_slice, res));
+        } else {
+            // Only take the write lock if this is the first time we've seen
+            // this blockhash in this slot.
+            let (_key_index, txs) = &mut *fork_entry
+                .entry(*blockhash)
+                .or_insert_with(|| (key_index, ConcurrentVec::new()));
+            txs.push((key_slice, res));
+        };
+    }
+}
+
+// Wrapper around boxcar::Vec that implements Serialize and Deserialize.
+#[cfg_attr(feature = "frozen-abi", derive(AbiExample))]
+#[derive(Debug)]
+pub struct ConcurrentVec<T> {
+    vec: boxcar::Vec<T>,
+}
+
+impl<T> ConcurrentVec<T> {
+    fn new() -> Self {
+        Self {
+            vec: boxcar::Vec::new(),
+        }
+    }
+
+    fn push(&self, item: T) {
+        self.vec.push(item);
+    }
+
+    fn iter(&self) -> impl Iterator<Item = (usize, &T)> {
+        self.vec.iter()
+    }
+}
+
+impl<T> IntoIterator for ConcurrentVec<T> {
+    type Item = T;
+    type IntoIter = boxcar::IntoIter<T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.vec.into_iter()
+    }
+}
+
+impl<'a, T> IntoIterator for &'a ConcurrentVec<T> {
+    type Item = (usize, &'a T);
+    type IntoIter = boxcar::Iter<'a, T>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.vec.iter()
+    }
+}
```

Comment:
maybe excepting serialize and deserialize, can't we just implement `Deref` to get access to the inner for all of this other stuff?